pkg/steps/template.go

```diff
@@ -421,24 +421,78 @@ func createOrRestartPod(podClient coreclientset.PodInterface, pod *coreapi.Pod)
 	return created, nil
 }
 
-func waitForPodDeletion(podClient coreclientset.PodInterface, name string, uid types.UID) error {
-	timeout := 300
-	for i := 0; i < timeout; i += 2 {
-		pod, err := podClient.Get(name, meta.GetOptions{})
-		if errors.IsNotFound(err) {
-			return nil
-		}
+func waitForTeardownToComplete(podClient coreclientset.PodInterface, name string, uid types.UID) error {
+	timeout := 5 * time.Minute
+
+	log.Printf("Waiting for pod %s to complete teardown ...", name)
+	wait.Poll(10*time.Second, timeout, func() (done bool, err error) {
+		pod, err := checkPodExistsAndValid(podClient, name, uid)
 		if err != nil {
-			return fmt.Errorf("could not retrieve deleting pod: %v", err)
+			// Infra error - failed to fetch pod data
+			return true, err
 		}
-		if pod.UID != uid {
-			return nil
+		if pod == nil {
+			// Pod was destroyed
+			return true, nil
 		}
-		log.Printf("Waiting for pod %s to be deleted ... (%ds/%d)", name, i, timeout)
-		time.Sleep(2 * time.Second)
+
+		for _, status := range append(append([]coreapi.ContainerStatus{}, pod.Status.InitContainerStatuses...), pod.Status.ContainerStatuses...) {
+			if status.Name == "teardown" && status.State.Terminated != nil {
+				// Teardown container finished
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	return nil
+}
+
+// Check that pod with this name exists and has the same UID
+func checkPodExistsAndValid(podClient coreclientset.PodInterface, name string, uid types.UID) (*coreapi.Pod, error) {
+	pod, err := podClient.Get(name, meta.GetOptions{})
+	switch {
+	case errors.IsNotFound(err), pod.UID != uid:
+		// Pod was removed or mismatching UID
+		return nil, nil
+	case err != nil:
+		return nil, fmt.Errorf("could not retrieve deleting pod: %v", err)
+	}
+	return pod, nil
+}
+
+func waitForPodDeletion(podClient coreclientset.PodInterface, name string, uid types.UID) error {
```

Comment:
`waitForPodDeletion` is no longer valid as a name -- were all callers expecting this new behavior?